src/java.base/share/classes/sun/security/provider/SHA3.java

```diff
@@ -103,48 +119,137 @@ private void implCompress0(byte[] b, int ofs) {
+        int numLongs = numBytes / 8;
 
-        for (int i = 0; i < numLongs - 1; i++) {
+        for (int i = 0; i < numLongs; i++) {
             asLittleEndian.set(out, ofs, state[i]);
             ofs += 8;
         }
-        if (numBytes == numLongs * 8) {
-            asLittleEndian.set(out, ofs, state[numLongs - 1]);
-        } else {
-            asLittleEndian.set(byteState, 0, state[numLongs - 1]);
-            System.arraycopy(byteState, 0,
-                    out, ofs, numBytes - (numLongs - 1) * 8);
+        if (numBytes % 8 != 0) {
+            asLittleEndian.set(byteState, 0, state[numLongs]);
+            System.arraycopy(byteState, 0, out, ofs, numBytes % 8);
+        }
+    }
+
+    void implSqueeze(byte[]output, int offset, int numBytes) {
+        // Moving this allocation to the block where t is used causes a little
+        // performance drop, that is why it is here.
+        byte[] byteState = new byte[8];
+        if (engineGetDigestLength() != 0) {
+            // This is not an XOF, so the squeeze() call is illegal.
+            throw new ProviderException("Squeezing is only allowed in XOF mode.");
+        }
+
+        if (squeezeOffset == -1) {
+            finishAbsorb();
+            squeezeOffset = 0;
+        }
+
+        int availableBytes = blockSize - squeezeOffset;
+
+        while (numBytes > availableBytes) {
+            int longOffset = squeezeOffset / 8;
+            int bytesToCopy = 0;
+
+            if (longOffset * 8 < squeezeOffset) {
+                asLittleEndian.set(byteState, 0, state[longOffset]);
+                longOffset++;
+                bytesToCopy = longOffset * 8 - squeezeOffset;
+                System.arraycopy(byteState, 8 - bytesToCopy,
+                        output, offset, bytesToCopy);
+                offset += bytesToCopy;
+            }
+            for (int i = longOffset; i < blockSize / 8; i++) {
+                asLittleEndian.set(output, offset, state[i]);
+                offset += 8;
+            }
+            keccak();
+            squeezeOffset = 0;
+            numBytes -= availableBytes;
+            availableBytes = blockSize;
+        }
+        // now numBytes <= availableBytes
+        int longOffset = squeezeOffset / 8;
+
+        if (longOffset * 8 < squeezeOffset) {
+            asLittleEndian.set(byteState, 0, state[longOffset]);
+            int bytesToCopy = min((longOffset + 1) * 8 - squeezeOffset, numBytes);
+            System.arraycopy(byteState, squeezeOffset - 8 * longOffset,
+                    output, offset, bytesToCopy);
+            longOffset++;
+            numBytes -= bytesToCopy;
+            offset += bytesToCopy;
+            squeezeOffset += bytesToCopy;
+
+            if (numBytes == 0) return;
+        }
+
+        int longsToConvert =
+                ((squeezeOffset + numBytes - longOffset * 8) + 7) / 8;
+
+        int limit = longOffset + longsToConvert - (numBytes % 8 == 0 ? 0 : 1);
```

Comment:
Can we also do the same simplification here as in `implDigest`?
```
        int longsToConvert =
                (squeezeOffset + numBytes - longOffset * 8) / 8;

        int limit = longOffset + longsToConvert;
```

Reply:
Fixed as suggested.